### ISSUE TITLE
Device registration failed with Device Provisioning Service

### ISSUE THREAD

**Issue by Vijay-Karthick** [Azure/azure-iot-sdk-python#189]
Hi,

I'm having trouble getting DPS to register the "riot-device-cert" device with my IoT Hub instance. I'm using a custom CA and have already validated it in both IoT Hub and DPS.

1. Using the riot-device-cert certificate and private key in the [https://github.com/Azure/azure-iot-sdk-python/blob/master/device/samples/iothub_client_sample_x509.py](url) example, I was able to successfully communicate with a manually added device (called "riot-device-cert") in my IoT Hub.
2. Next, I deleted this device and created an individual enrollment in Device Provisioning Service by uploading the same device certificate from above. PS: This DPS is already linked to the previous IoT Hub.
3. I tried running the example code in [https://github.com/Azure/azure-iot-sdk-python/blob/master/provisioning_device_client/samples/provisioning_device_client_sample.py](url) and the device registration failed with a message "Invalid certificate". I've mentioned the output below for review.

Python 2.7.12 (default, Dec  4 2017, 14:50:18)
[GCC 5.4.0 20160609]
Provisioning Device Client for Python
Starting the Provisioning Client Python sample...
Scope ID=0ne00026980
Security Device Type X509
Protocol HTTP

Provisioning API Version: 1.2.8

Press Enter to interrupt...

Register status callback:
reg_status = CONNECTED
user_context = None

PUT /0ne00026980/registrations/riot-device-cert/register?api-version=2017-11-15 HTTP/1.1
UserAgent: prov_device_client/1.0
Accept: application/json
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Host: global.azure-devices-provisioning.net:443
content-length: 39

len: 39
{ "registrationId":"riot-device-cert" }

HTTP Status: 202

date: Thu, 06 Sep 2018 14:51:24 GMT
content-type: application/json; charset=utf-8
transfer-encoding: chunked
location: https://global.azure-devices-provisioning.net:443/0ne00026980/registrations/riot-devicecert/register
x-ms-request-id: 0cc4723c-233c-4f67-a903-d91122f6e8cb
strict-transport-security: max-age=31536000; includeSubDomains

{"operationId":"4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8","status":"assigning"}

Register status callback:
reg_status = ASSIGNING
user_context = None

GET /0ne00026980/registrations/riot-device-cert/operations/4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8?api-version=2017-11-15 HTTP/1.1
UserAgent: prov_device_client/1.0
Accept: application/json
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Host: global.azure-devices-provisioning.net:443
content-length: 0

HTTP Status: 202

date: Thu, 06 Sep 2018 14:51:24 GMT
content-type: application/json; charset=utf-8
transfer-encoding: chunked
location: https://global.azure-devices-provisioning.net:443/0ne00026980/registrations/riot-device-cert/operations/4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8
retry-after: 3
x-ms-request-id: f7e04af5-a08d-4305-b604-641d3a1ccc80
strict-transport-security: max-age=31536000; includeSubDomains

{"operationId":"4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8","status":"assigning","registrationState":{"registrationId":"riot-device-cert","status":"assigning"}}

Register status callback:
reg_status = ASSIGNING
user_context = None

GET /0ne00026980/registrations/riot-device-cert/operations/4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8?api-version=2017-11-15 HTTP/1.1
UserAgent: prov_device_client/1.0
Accept: application/json
Connection: keep-alive
Content-Type: application/json; charset=utf-8
Host: global.azure-devices-provisioning.net:443
content-length: 0

HTTP Status: 200

date: Thu, 06 Sep 2018 14:51:27 GMT
content-type: application/json; charset=utf-8
transfer-encoding: chunked
x-ms-request-id: 8bc7def0-64a7-418e-ae1d-73ff3d1788d2
strict-transport-security: max-age=31536000; includeSubDomains

{"operationId":"4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8","status":"failed","registrationState":{"registrationId":"riot-device-cert","createdDateTimeUtc":"2018-09-06T14:51:25.7932051Z","status":"failed","errorCode":401002,"errorMessage":"Invalid certificate","lastUpdatedDateTimeUtc":"2018-09-06T14:51:26.0101229Z","etag":"\u00225300215f-0000-0000-0000-5b913eee0000\u0022"}}

Error: Time:Thu Sep  6 10:51:28 2018 File:/usr/sdk/src/c/provisioning_client/src/prov_device_ll_client.c Func:prov_transport_process_json_reply Line:407 Provisioning Failure: OperationId: 4.87f7851aaaa18ed1.90f91c73-6e65-49c5-b7c8-d5b6b38688b8 - Date: 2018-09-06T14:51:26.0101229Z - Msg: Invalid certificate
Error: Time:Thu Sep  6 10:51:28 2018 File:/usr/sdk/src/c/provisioning_client/src/prov_transport_http_client.c Func:prov_transport_http_dowork Line:915 Unable to process registration reply.
Error: Time:Thu Sep  6 10:51:28 2018 File:/usr/sdk/src/c/provisioning_client/src/prov_device_ll_client.c Func:on_transport_registration_data Line:493 Failure retrieving data from the provisioning service

Register device callback:
    register_result = DEV_AUTH_ERROR
    iothub_uri = None
    user_context = None

Device registration failed!

Thanks,
Vijay.

**Comment by pierreca** [comment]
The `riot-device-cert` name is probably the common name of your certificate and is used as the registration id by the device provisioning service. the desired device id must be entered in the Device Id field when you create the individual enrollment object on the Device Provisioning Service and will be assigned when the registration finishes.

I'm not sure why the registration fails with this error but I'll ask the C SDK folks what the error means exactly (ping @jebrando) 

Is that certificate self-signed? or is it signed by a CA? are you using individual enrollments or group enrollments? if with group enrollments, have you completed the proof of possession step?

**Comment by Vijay-Karthick** [comment]
Hi @pierreca , @jebrando 

Thanks. The device is signed by CA. It is added using individual enrollment. Proof of possession is already completed for both IoT Hub as well as device provisioning service.

**Comment by Vijay-Karthick** [comment]
Hi @pierreca , @jebrando , 

Tagging you again. 

@zolvarga : Can you please look into this too?

**Comment by pierreca** [comment]
@Vijay-Karthick do you have an intermediate certificate and a leaf certificate? are you passing down the whole cert chain?

**Comment by pierreca** [comment]
related to #272 (and answered there)
TL;DR: fixed in v2

**Comment by az-iot-builder-01** [comment]
@Vijay-Karthick, @anhashia, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
